### PR TITLE
signup route (complete signup)

### PR DESCRIPTION
**Pull request by MohammadKoolivand**
completed signup finally

### CODE REVIEW

File: src/controllers/app/userController.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ exports.SignUp = function (req, res) {
       }
       newUser.save((err, savedUser) => {
         if (err) {
-          return res.status(500).json({ message: "Server Error" });
+          return res.status(500).json({ message: "Server Error 523" });
```
Comment (farshid0594) [inline src/controllers/app/userController.js]: server error 523??

File: src/controllers/app/userController.js [hunk 2]
```diff
@@ -95,3 +90,42 @@ exports.CheckCode = function (req, res) {
     }
   });
 };
+
+exports.completeSignup = function (req, res) {
+  const { operationId, userName, password } = req.body;
+  var { userId, code } = jwt.verify(operationId, process.env.JWT_SECRET_KEY);
```
Comment (farshid0594) [inline src/controllers/app/userController.js]: change var to const please please please please!!!!!!!!!!

File: src/controllers/app/userController.js [hunk 3]
```diff
@@ -95,3 +90,42 @@ exports.CheckCode = function (req, res) {
     }
   });
 };
+
+exports.completeSignup = function (req, res) {
+  const { operationId, userName, password } = req.body;
+  var { userId, code } = jwt.verify(operationId, process.env.JWT_SECRET_KEY);
+  console.log("userId", userId);
```
Comment (farshid0594) [inline src/controllers/app/userController.js]: remove log

File: src/controllers/app/userController.js [hunk 4]
```diff
@@ -95,3 +90,42 @@ exports.CheckCode = function (req, res) {
...
+        return res.status(404).json({ message: "user not found" });
+      } else {
+        if (foundedUser.code.text !== code) {
+          return res.status(400).json({ message: "wrong enteries" });
+        }
+        if (foundedUser.isActive) {
+          return res.status(400).json({ message: "user is already active" });
+        }
+        //updated inActive user
+        updatedUser = foundedUser;
```
Comment (farshid0594) [inline src/controllers/app/userController.js]: Where did you define updatedUser variable?

File: src/controllers/app/userController.js [hunk 5]
```diff
@@ -95,3 +90,42 @@ exports.CheckCode = function (req, res) {
...
+        updatedUser.username = userName;
+        updatedUser.password = password;
```
Comment (farshid0594) [inline src/controllers/app/userController.js]: Where did you hash  password?

File: src/controllers/app/userController.js [hunk 6]
```diff
@@ -95,3 +90,42 @@ exports.CheckCode = function (req, res) {
...
+        updatedUser.isActive = true;
```
Comment (farshid0594) [inline src/controllers/app/userController.js]: Where did you check that the username is not duplicate?

File: src/controllers/app/userController.js [hunk 7]
```diff
@@ -95,3 +90,42 @@ exports.CheckCode = function (req, res) {
...
+        updatedUser.save((err, savedUpdatedUser) => {
+          if (err) {
+            return res.status(500).json({ message: "Server Error" });
+          }
+          return res.status(200).json({
+            message: "signup completed",
+            id: savedUpdatedUser._id,
```
Comment (farshid0594) [inline src/controllers/app/userController.js]: Why did you return the ID?

File: src/validations/loginValidate.js [hunk 8]
```diff
@@ -8,3 +8,26 @@ exports.ValidateMobile = (req, res, next) => {
...
+      password: "password length must be more than 6",
+    });
+  }
+  if (userName.length < 4) {
+    errors.push({
+      key: "userName",
+      userName: "userName length must be more than 6",
+    });
+  }
+
```
Comment (farshid0594) [inline src/validations/loginValidate.js:29]: Check that the username is unique

File: src/validations/loginValidate.js [hunk 9]
```diff
@@ -8,3 +8,26 @@ exports.ValidateMobile = (req, res, next) => {
     next();
   }
 };
+
+
+exports.ValidateSignupFields = (req, res, next) => {
+  var errors = [];
+  var { userName, password } = req.body;
```
Comment (farshid0594) [inline src/validations/loginValidate.js]: change var to const in lines [14,15]

File: src/controllers/app/userController.js (hunk 1 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js]: done

File: src/controllers/app/userController.js (hunk 2 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js]: done

File: src/controllers/app/userController.js (hunk 3 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js]: done

File: src/controllers/app/userController.js (hunk 7 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js]: done

File: src/validations/loginValidate.js (hunk 9 above)
Comment (MohammadKoolivand) [reply src/validations/loginValidate.js]: done

File: src/controllers/app/userController.js (hunk 4 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js]: done

File: src/controllers/app/userController.js (hunk 6 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js]: done

File: src/controllers/app/userController.js (hunk 5 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js]: done

File: src/validations/loginValidate.js (hunk 8 above)
Comment (MohammadKoolivand) [reply src/validations/loginValidate.js:29]: done

**Review by amirhnr — APPROVED** [review]

File: src/validations/loginValidate.js [hunk 10]
```diff
@@ -29,5 +30,17 @@ exports.ValidateSignupFields = (req, res, next) => {
...
+    if (doc !== null) {
+      errors.push({
+        key: "userName",
+        password: "userName exists",
+      });
+      return res.status(400).json({ errors: errors });
+    }
+    next();
+  });
 };
```
Comment (farshid0594) [inline src/validations/loginValidate.js:45]: replace lines: 30 - 45  with:


  User.findOne({ userName: userName }, (err, doc) => {
    if (err) {
      return res.status(500).json({ message: "server Error"});
    }
    if (doc !== null) {
      errors.push({
        key: "userName",
        password: "userName exists",
      });
    }
 if (errors.length > 0) {
    return res.status(400).json({ errors: errors });
  }
    next();
  });

**Review by farshid0594 — APPROVED** [review]
Goooooood!

File: src/validations/loginValidate.js (hunk 10 above)
Comment (MohammadKoolivand) [reply src/validations/loginValidate.js:45]: done

**Pull request merged**